ambari-server/src/main/java/org/apache/ambari/server/upgrade/UpgradeCatalog300.java

```diff
@@ -383,4 +389,94 @@ protected void updateKerberosConfigurations() throws AmbariException {
       }
     }
   }
+
+  /**
+   * Moves LDAP related properties from ambari.properties to ambari_configuration DB table
+   * @throws AmbariException if there was any issue when clearing ambari.properties
+   */
+  protected void upgradeLdapConfiguration() throws AmbariException {
```

Comment:
Upgrade logic should be idempotent:

1. Running upgrade on already upgraded setup should be no-op (no errors etc.).  I guess this is the case, since properties removed at the end will not be processed on next try.
2. Retrying partially completed upgrade should resume without errors.  This _might_ be a  problem here, since all properties are left until the end, duplicate properties may be created.  Can you please confirm if this handled correctly?  (Test by throwing some exception in the middle of property processing.)  If not, can you please make sure to "create or update" the properties instead of "create".

Reply:
These are good notices; thanks for bringing them up.
Point 1: like you indicated there won't be more properties to be upgraded = no-op
Point 2: I'll check it out and modify the code if needed

Reply:
Fixed